src/sources/envoy_als/grpc.rs

```diff
@@ -0,0 +1,100 @@
+use crate::{
+    internal_events::{EventsReceived, StreamClosedError},
+    shutdown::ShutdownSignal,
+    SourceSender,
+};
+use envoy_proto::envoy::service::accesslog::v3::{
+    access_log_service_server::AccessLogService, stream_access_logs_message::LogEntries,
+    StreamAccessLogsMessage, StreamAccessLogsResponse,
+};
+use tonic::{Request, Response, Status, Streaming};
+use value::Value;
+use vector_common::internal_event::{CountByteSize, InternalEventHandle as _, Registered};
+use vector_core::{event::LogEvent, EstimatedJsonEncodedSizeOf};
+
+#[derive(Clone)]
+pub(super) struct Service {
+    pub events_received: Registered<EventsReceived>,
+    pub pipeline: SourceSender,
+    pub shutdown: ShutdownSignal,
+}
+
+#[tonic::async_trait]
+impl AccessLogService for Service {
+    async fn stream_access_logs(
+        &self,
+        request: Request<Streaming<StreamAccessLogsMessage>>,
+    ) -> Result<Response<StreamAccessLogsResponse>, Status> {
+        let mut in_stream = request.into_inner();
+        let mut shutdown = self.shutdown.clone();
+        let mut stream_identifier = None;
+
+        loop {
+            tokio::select! {
+                _ = &mut shutdown => break,
+                stream_msg = in_stream.message() => {
+                    match stream_msg {
+                        Ok(omsg) => {
+                            match omsg {
+                                Some(msg) => {
+                                    let mut events = vec![];
+
+                                    if let Some(logs_entries) = msg.log_entries {
+                                        match logs_entries {
+                                            LogEntries::HttpLogs(http_logs) => {
+                                                for l in http_logs.log_entry {
+                                                    let mut evt = LogEvent::default();
+                                                    evt.insert("http_log", Value::from(l));
+
+                                                    if let Some(identifier) = msg.identifier.clone() {
+                                                        stream_identifier = Some(identifier);
+                                                    }
+                                                    if let Some(stream_id) = stream_identifier.clone() {
+                                                        evt.insert("identifier", Value::from(stream_id));
+                                                    }
```

Comment:
This logic seems off 🤔 . is the clone of `stream_identifier` necessary? Seems like the second `if let Some` is not necessary. Looks like `stream_identifier` doesn't need to be defined outside of this scope.

Reply:
So the way this works is `identifier` is only send on the first message of an ALS stream. We want to have this data available on all logs however, so we store it in memory (on `stream_identifier`) when it's received. On subsequent messages, the `msg.identifier` will be `None` and we rely on the stored `stream_identifier` for that info.

It seemed safer to have that wrapped in an optional rather than just assume the `stream_identifier` is set. The clone is needed because the `stream_identifier` will be moved upon use.

I will add a comment in the code here explaining all this.

Reply:
Thanks for the additional context! I understand now the reasoning with stream_identifier and the comments/ helper functions help make that more obvious.

What I'm still wondering about- is it not possible to modify the `From` function in convert.rs , to accept a reference to the Identifier, so that you don't have to clone it? It seems like that should be do able because the from() function looks to be just referencing data from the Identifier.